### PR TITLE
fix: Support parameterized source code input for TrainingStep

### PR DESCRIPTION
**Pull request by qidewenwhen**
*Description of changes:* The [change in v2.75](https://github.com/aws/sagemaker-python-sdk/commit/219617ab16e66d7a71df4ce31e03615e408c847a) has broken the parameterization of source code input for TrainingStep. This PR fixes this issue and re-supports it.

*Testing done:* unit tests and integ tests

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/amazon/hyperparameter.py [hunk 1]
```diff
@@ -98,8 +100,14 @@ def serialize_all(obj):
         """
         if "_hyperparameters" not in dir(obj):
             return {}
-        return {
-            k: json.dumps(v) if isinstance(v, list) else str(v)
-            for k, v in obj._hyperparameters.items()
-            if v is not None
-        }
+        hps = {}
+        for k, v in obj._hyperparameters.items():
```
Comment (jerrypeng7773) [inline src/sagemaker/amazon/hyperparameter.py:104]: how this change is related to enable the `source_dir` parameterization?

File: src/sagemaker/fw_utils.py [hunk 2]
```diff
@@ -124,6 +125,43 @@ def validate_source_dir(script, directory):
...
+            raise TypeError(
+                "entry_point, source_dir should not be pipeline variables when git_config is given."
+            )
+    if is_pipeline_variable(entry_point):
+        if not source_dir:
+            raise TypeError(
+                "The entry_point should not be a pipeline variable when source_dir is missing."
+            )
+        logger.warning(
+            "The entry_point is a pipeline variable: %s. Its execution-time value has to be "
```
Comment (jerrypeng7773) [inline src/sagemaker/fw_utils.py:153]: its compile time?

File: src/sagemaker/jumpstart/utils.py [hunk 3]
```diff
@@ -271,26 +271,41 @@ def add_jumpstart_tags(
         training_script_uri (Optional[str]): S3 URI for training script tarball.
             (Default: None).
     """
-
+    warn_msg = (
+        "The uri (%s) is a pipeline variable which is parsed only in execution time. "
```
Comment (jerrypeng7773) [inline src/sagemaker/jumpstart/utils.py:275]: parsed --> resolved?

Comment (jerrypeng7773) [reply src/sagemaker/jumpstart/utils.py:275]: Thus skip adding custom tags to JumpStart models for this uri in SDK.

--->

Thus skip adding custom tags to JumpStart models for this uri during compile time. ?

File: tests/integ/sagemaker/workflow/test_training_steps.py [hunk 4]
```diff
@@ -67,11 +68,17 @@ def test_training_job_with_debugger_and_profiler(
...
     ]
     debugger_hook_config = DebuggerHookConfig(
-        s3_output_path=(f"s3://{sagemaker_session.default_bucket()}/{uuid.uuid4()}/tensors")
+        s3_output_path=f"s3://{sagemaker_session.default_bucket()}/{uuid.uuid4()}/tensors"
     )
 
     base_dir = os.path.join(DATA_DIR, "pytorch_mnist")
-    script_path = os.path.join(base_dir, "mnist.py")
+    entry_point = "mnist.py"
+    source_dir = sagemaker_session.upload_data(
```
Comment (jerrypeng7773) [inline tests/integ/sagemaker/workflow/test_training_steps.py:76]: are we using this variable `source_dir` somewhere ?

File: src/sagemaker/fw_utils.py [hunk 5]
```diff
@@ -124,6 +125,43 @@ def validate_source_dir(script, directory):
...
+            "a local path to a Python source file which located at the root of source_dir.",
```
Comment (staubhp) [inline src/sagemaker/fw_utils.py:154]: nit: "which is located"

File: src/sagemaker/jumpstart/utils.py (hunk 3 above)
Comment (staubhp) [reply src/sagemaker/jumpstart/utils.py:275]: These comments might seem nitty but let's treat the warnings as docs as far as readability. 

Suggested:
`The URI (%s) is a pipeline variable which is only parsed (interpreted?) at execution time. As a result, the JumpStart resources will not be tagged.`

File: src/sagemaker/fw_utils.py (hunk 2 above)
Comment (qidewenwhen) [reply src/sagemaker/fw_utils.py:153]: No, I put execution-time value here on purpose. The thing is eventually for training job to correctly work, the `entry_point` has to be a local path in the container env to a file which is located at the root of `source_dir`.
We're not able to validate this in compile time, so I add a warning to remind users.

File: src/sagemaker/jumpstart/utils.py (hunk 3 above)
Comment (qidewenwhen) [reply src/sagemaker/jumpstart/utils.py:275]: Good call out! Updated in the new commit.

File: src/sagemaker/amazon/hyperparameter.py (hunk 1 above)
Comment (qidewenwhen) [reply src/sagemaker/amazon/hyperparameter.py:104]: Good call.

Without this fix, the existing unit test will fail: `test_training_step_with_algorithm_base`

The `source_dir` and `entry_point` should be added into the hyperparameter, that's why this fix is required here/

**Review by staubhp — APPROVED** [review]

File: src/sagemaker/fw_utils.py (hunk 2 above)
Comment (jerrypeng7773) [reply src/sagemaker/fw_utils.py:153]: nit: 

ok, I found this description a bit confusing, how about:

> During pipeline execution, `entry_point` has to be resolved to a value pointing to a local path in the container, blah, blah.

**Review by jerrypeng7773 — APPROVED** [review]

File: src/sagemaker/fw_utils.py (hunk 2 above)
Comment (qidewenwhen) [reply src/sagemaker/fw_utils.py:153]: Almost missed this comment. Making sense. Updated in the new commit.

Comment (qidewenwhen) [reply src/sagemaker/fw_utils.py:153]: Updated to:
```
        logger.warning(
            "The entry_point is a pipeline variable: %s. During pipeline execution, "
            "the interpreted value of entry_point has to be a local path in the container, "
            "pointing to a Python source file which is located at the root of source_dir.",
            type(entry_point),
        )
   ```

File: src/sagemaker/fw_utils.py [hunk 6]
```diff
@@ -124,6 +125,43 @@ def validate_source_dir(script, directory):
...
+            type(entry_point),
+        )
+    if is_pipeline_variable(source_dir):
+        logger.warning(
+            "The source_dir is a pipeline variable: %s. Its execution-time value has to be "
```
[the remark this answers is not in the document]
Comment (qidewenwhen) [reply src/sagemaker/fw_utils.py:159]: Updated to
```
logger.warning(
            "The source_dir is a pipeline variable: %s. During pipeline execution, "
            "the interpreted value of source_dir has to be an S3 URI and "
            "must point to a tar.gz file",
            type(source_dir),
        )
```

**Review by jerrypeng7773 — APPROVED** [review]

**Review by navaj0 — APPROVED** [review]

**Review by staubhp — APPROVED** [review]

**Review by BasilBeirouti — APPROVED** [review]

**Pull request merged**